src/Compression/CompressionCodecFSST.cpp

```diff
@@ -0,0 +1,144 @@
+                encoder,
+                rows_count,
+                len_in.data(),
+                str_in.data(),
+                out_size,
+                reinterpret_cast<unsigned char *>(dest + header_size),
+                len_out,
+                str_out)
+            < rows_count)
+            throw std::runtime_error("FSST compression failed");
+        /* fsst_destroy(encoder); TODO(ebek): Понять почему вызывается деструктор какого-то левого кодека */
+
+        /* Copy prerequisites to dest */
+        memcpy(dest + fsst_header_size, &rows_count, sizeof(rows_count));
+        memcpy(dest + fsst_header_size + sizeof(rows_count), len_out, sizeof(len_out));
+        memcpy(dest + fsst_header_size + sizeof(rows_count) + sizeof(len_out), len_in.data(), len_in.size() * sizeof(size_t));
+
+        /* Count data total compressed size without header */
+        UInt32 compressed_size{0};
+        for (size_t i = 0; i < rows_count; ++i)
+            compressed_size += len_out[i];
+        return static_cast<UInt32>(header_size) + compressed_size;
+    }
+
+    void doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size) const override
+    {
+        UNUSED(uncompressed_size, source_size);
+        fsst_decoder_t decoder;
+        size_t fsst_header_size = fsst_import(&decoder, reinterpret_cast<unsigned char *>(const_cast<char *>(source)));
+
+        size_t rows_count;
+        memcpy(&rows_count, source + fsst_header_size, sizeof(rows_count));
+
+        size_t lens[rows_count];
+        std::vector<size_t> len_in(rows_count);
+        memcpy(lens, source + fsst_header_size + sizeof(rows_count), sizeof(lens));
+        memcpy(len_in.data(), source + fsst_header_size + sizeof(rows_count) + sizeof(lens), sizeof(size_t) * len_in.size());
+        size_t header_size{fsst_header_size + sizeof(rows_count) + sizeof(lens) + sizeof(size_t) * len_in.size()};
+
+        const char * str{source + header_size};
+        for (size_t i = 0; i < rows_count; ++i)
+        {
+            dest = writeVarUInt(len_in[i], dest);
+
+            auto decompressed_size = fsst_decompress(
+                &decoder,
+                lens[i],
+                reinterpret_cast<const unsigned char *>(str),
+                out_size,
+                reinterpret_cast<unsigned char *>(dest));
+
+            str += lens[i];
+            dest += decompressed_size;
+        }
+    }
+
+    UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override { return uncompressed_size + FSST_MAXHEADER; }
+
+    bool isCompression() const override { return true; }
+    bool isGenericCompression() const override { return true; }
+
+private:
+    void splitDataByRows(
+        const unsigned char * data, std::vector<const unsigned char *> & rows, std::vector<size_t> & lens, UInt32 source_size) const
+    {
+        const unsigned char * end{data + source_size};
+
+        while (data != end)
+        {
+            UInt64 cur_len;
```

Comment:
I am pleasantly surprised that ClickHouse serializes String columns as is consecutive strings instead as two arrays (strings + length). That, of course, simplifies the job of `splitDataByRows`.